### DISCUSSION TITLE
Optimizing memory usage

### DISCUSSION THREAD

**Discussion by MattIPv4** [discordjs/discord.js#5250] [General]
Hey folks,

Wanted to create a discussion/thread where folks can share ideas on the best ways to optimize and reduce memory usage in Discord.js. I want to focus mostly on ways to reduce the memory usage that Discord.js itself brings, versus what a user might add in with their own data held in memory.

For me, I use the built-in message sweeping with the following options:

```js
messageCacheMaxSize: 10,
messageCacheLifetime: 5 * 60,
messageSweepInterval: 5 * 60,
```

I also have custom sweeping set up on users in the bot, to help reduce memory usage further:

```js
userCacheLifetime: 60 * 60,
userSweepInterval: 5 * 60,

...

    client.purgeCachedUsers = () => {
        const purgableUser = user => {
            if (user.id === client.user.id) return false;
            if (!user.lastMessageID) return true;
            const flake = SnowflakeUtil.deconstruct(user.lastMessageID);
            const diff = Date.now() - flake.timestamp;
            return (diff / 1000 > client.options.userCacheLifetime);
        };

        client.users.cache.each(user => {
            if (purgableUser(user))
                client.users.cache.delete(user.id);
        });

        client.guilds.cache.each(guild => {
            guild.members.cache.each(member => {
                if (purgableUser(member))
                    guild.members.cache.delete(member.id);
            });
        });
    };
```

What tips/tricks do y'all have for keeping memory usage down?

**Comment by kanetjuh** [comment]
less code and files

**Comment by kyranet** [comment]
This should be solved with the new [cache customization](https://discordjs.guide/miscellaneous/cache-customization.html), so I'll lock the discussion as resolved.
